### PR TITLE
Run `OtaDownloader` as a separate thread

### PR DESCRIPTION
**Pull request by de-sh**
Move code associated with OTAs to run on a separate thread.

### CODE REVIEW

File: uplink/src/base/actions/ota.rs [hunk 1]
```diff
@@ -240,10 +240,67 @@ impl OtaDownloader {
...
+        let mut expected_forward = ota_update.clone();
+        expected_forward.ota_path = Some(ota_path);
+        let ota_action = Action {
+            action_id: "1".to_string(),
+            kind: "firmware_update".to_string(),
+            name: "firmware_update".to_string(),
+            payload: serde_json::json!(ota_update).to_string(),
+        };
+
+        // Send action to OtaDownloader with OtaTx
```
[the remark this answers is not in the document]
Comment (de-sh) [reply uplink/src/base/actions/ota.rs:292]: Fixed it, the issue was with how streams were being initialized

**Review by de-sh — COMMENTED** [review]
Finished adding unit tests

**Comment by tekjar** [comment]
I think we can do a few clever things to significantly simplify code here. Let's discuss this tomorrow

File: uplink/src/base/actions/ota.rs [hunk 2]
```diff
@@ -13,97 +57,179 @@ use crate::base::{Config, Stream};
...
-        // Create handler to perform download from URL
-        let resp = client.get(url).send().await?;
+        // Ensure that directory for downloading file into, of the format `path/to/{version}/`, exists
+        let mut ota_path = PathBuf::from(self.config.ota.path.clone());
+        ota_path.push(&update.version);
+        create_dir_all(&ota_path)?;
+
+        // Create file to actually download into
+        let mut file_path = ota_path.to_owned();
+        let file_name = url.split("/").last().ok_or(Error::FileNameMissing(url.to_owned()))?;
```
Comment (tekjar) [inline uplink/src/base/actions/ota.rs:177]: Is error handling necessary here given we are creating the directory and erroring out if we can't. When does this call return `None`?

Comment (tekjar) [reply uplink/src/base/actions/ota.rs:177]: Ah we are creating the same file name as image name in url path

Comment (tekjar) [reply uplink/src/base/actions/ota.rs:177]: Why not extract all of file creation into a method like the original branch?

**Pull request merged**